### PR TITLE
New package: py-qiskit-ibm-provider and more

### PR DESCRIPTION
**Pull request by qwertos**
- py-singledispatchmethod: New package
- py-rustworkx: New package
- py-ipyvue: New package
- py-websocket-client: Added version 1.5.1
- py-contextvars: New package
- py-ipyvuetify: New package
- py-qiskit-terra: Added version 0.23.3
- py-qiskit-ibm-provider: New package

### CODE REVIEW

**Comment by qwertos** [comment]
@spackbot fix style

**Review by tldahlgren — COMMENTED** [review]
Confirmed the 7 version sha256.

**Comment by tldahlgren** [comment]
Deferring python dependency assessment and  merge decision to @adamjstewart and or @pradyunsg

File: var/spack/repos/builtin/packages/py-contextvars/package.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+class PyContextvars(PythonPackage):
+    """This package implements a backport of Python 3.7 contextvars module (see
+    PEP 567) for Python 3.6."""
+
+    homepage = "http://github.com/MagicStack/contextvars"
+    pypi = "contextvars/contextvars-2.4.tar.gz"
+
+    version("2.4", sha256="f38c908aaa59c14335eeea12abea5f443646216c4e29380d7bf34d2018e2c39e")
+
+    depends_on("python@3.6:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-contextvars/package.py]: Where do you see this? I would just remove this line, Spack only supports 3.7+ anyway

File: var/spack/repos/builtin/packages/py-ipyvue/package.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class PyIpyvue(PythonPackage):
+    """Jupyter widgets base for Vue libraries"""
+
+    homepage = "https://github.com/mariobuikhuizen/ipyvue"
+    pypi = "ipyvue/ipyvue-1.9.0.tar.gz"
+
+    version("1.9.0", sha256="841a8dbe0e9d2b1e8fe58a35ee7500f73b66bd08f3f8284d1569c3d8e38fa775")
+
+    depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipyvue/package.py:17]: Missing ipywidgets dep. Also, is npm required to build?

File: var/spack/repos/builtin/packages/py-ipyvuetify/package.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+class PyIpyvuetify(PythonPackage):
+    """Jupyter widgets based on vuetify UI components"""
+
+    homepage = "https://github.com/widgetti/ipyvuetify"
+    pypi = "ipyvuetify/ipyvuetify-1.8.5.tar.gz"
+
+    version("1.8.5", sha256="c1f88485dcac8324c8e134cf38fcea1fb5f9bd3e75836d055713e354f158f565")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-ipyvue@1.5:2", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipyvuetify/package.py:18]:
```suggestion
    depends_on("py-ipyvue@1.5:1", type=("build", "run"))
```
Is npm required to build?

File: var/spack/repos/builtin/packages/py-qiskit-ibm-provider/package.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+from spack.package import *
+
+
+class PyQiskitIbmProvider(PythonPackage):
+    """This project contains a provider that allows accessing the IBM Quantum
+    systems and simulators."""
+
+    homepage = "https://github.com/Qiskit/qiskit-ibm-provider"
+    url = "https://github.com/Qiskit/qiskit-ibm-provider/archive/refs/tags/0.4.0.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-ibm-provider/package.py:14]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-qiskit-terra/package.py [hunk 5]
```diff
@@ -47,3 +48,13 @@ class PyQiskitTerra(PythonPackage):
...
+
+    with when("@0.23.3"):
+        depends_on("python@3.7:", type=("build", "run"))
+        depends_on("py-setuptools-rust", type="build")
+        depends_on("py-rustworkx@0.12.0:", type=("build", "run"))
+        depends_on("py-scipy@1.5:", type=("build", "run"))
+        depends_on("py-symengine@0.9:", type=("build", "run"))
+        depends_on("py-typing-extensions", when="^python@:3.7", type=("build", "run"))
+        depends_on("py-shared-memory38", when="^python@:3.7", type=("build", "run"))
+        depends_on("py-importlib-metadata@:4", when="^python@:3.7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-terra/package.py:61]: Missing singledispatch dep

File: var/spack/repos/builtin/packages/py-qiskit-terra/package.py [hunk 6]
```diff
@@ -15,6 +15,7 @@ class PyQiskitTerra(PythonPackage):
     homepage = "https://github.com/Qiskit/qiskit-terra"
     pypi = "qiskit-terra/qiskit-terra-0.18.3.tar.gz"
 
+    version("0.23.3", sha256="8f2f61622c28ad38db95d9bbacb03b9743923581545df08eb5047fa86f4de198")
     version("0.22.2", sha256="efd212cd98479ebedc8cc1f93d4eb8039f21c07bd39a62065b584e02d72e632d")
     version("0.18.3", sha256="8737c8f1f4c6f29ec2fb02d73023f4854a396c33f78f4629a861a3e48fc789cc")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-terra/package.py:21]: tweedledum is no longer required

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by qwertos** [comment]
@spackbot fix style

**Comment by qwertos** [comment]
I'm going to simplify and resubmit

**Pull request closed without merging**